### ISSUE TITLE
[bug] LLVM IR Fixed array values are not returned correctly if larger than 64 bytes when using -O0 on arm64

### ISSUE THREAD

**Issue by IsaacShelton** [llvm/llvm-project#56295]
### Details

**Description:**  When lowering, LLVM IR Fixed array values are not returned correctly if larger than 64 bytes when using `-O0` on arm64
**Target:** arm64-apple-darwin21.4.0
**Version:** LLVM 14.0.6 and earlier (until at least LLVM 13.0.1)

### Abstract

Returning a fixed array value doesn't not seem to be correctly lowered when using `-O0` on arm64 if the array is larger than 64 bytes.

This issue isn't for any particular language, but I found it using the following code:

```
import 'sys/cstdio.adept'

func getValues() 9 long {
    values 9 long
    
    repeat 9 {
        values[idx] = idx as long
    }
    
    return values
}

func main {
    values 9 long = getValues()
    
    repeat 9 {
        printf('%c\n', 'A'ub + values[idx] as ubyte)
    }
}
```

When the fixed arrays have 8 items (or equivalent 64 bytes), it works correctly. But as soon as you increase it to 9 (or any fixed array with >64 bytes), the LLVM IR is not lowered correctly anymore (at least for `-O0` on arm64).

To demonstrate this, here are two LLVM IR files. The only difference between them, is that the size of the fixed arrays and iteration counts are increased from 8 to 9.

Both compile correctly when using `-O1`, but the `error.ll` does not when using `-O0`.

The expected output for both programs looks like this: (but `I` is only for `error.ll` since it shows 9 characters instead of 8):
```console
A
B
C
D
E
F
G
H
I
```

However, when compiling `error.ll` with `-O0` you get:
```console
A
A
A
A
A
A
A
y
A
```

### Complete LLVM IR

`ok.ll`
```llvm
source_filename = ""
target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128"
target triple = "arm64-apple-darwin21.4.0"

@S0 = internal constant [4 x i8] c"%c\0A\00"

declare i32 @printf(i8* %0, ...)

; Function Attrs: nounwind
define private [8 x i64] @a41() #0 {
  %1 = alloca [8 x i64], align 8
  %2 = alloca i64, align 8
  store [8 x i64] zeroinitializer, [8 x i64]* %1, align 8
  store i64 0, i64* %2, align 8
  br label %13

3:                                                ; preds = %13
  %4 = load i64, i64* %2, align 8
  %5 = load i64, i64* %2, align 8
  %6 = bitcast [8 x i64]* %1 to i64*
  %7 = getelementptr i64, i64* %6, i64 %5
  store i64 %4, i64* %7, align 8
  br label %8

8:                                                ; preds = %3
  %9 = load i64, i64* %2, align 8
  %10 = add i64 %9, 1
  store i64 %10, i64* %2, align 8
  br label %13

11:                                               ; preds = %13
  %12 = load [8 x i64], [8 x i64]* %1, align 8
  ret [8 x i64] %12

13:                                               ; preds = %8, %0
  %14 = load i64, i64* %2, align 8
  %15 = icmp ult i64 %14, 8
  br i1 %15, label %3, label %11
}

; Function Attrs: nounwind
define i32 @main() #0 {
  br label %20

1:                                                ; preds = %20
  %2 = alloca [8 x i64], align 8
  %3 = alloca i64, align 8
  %4 = call [8 x i64] @a41()
  store [8 x i64] %4, [8 x i64]* %2, align 8
  store i64 0, i64* %3, align 8
  br label %17

5:                                                ; preds = %17
  %6 = load i64, i64* %3, align 8
  %7 = bitcast [8 x i64]* %2 to i64*
  %8 = getelementptr i64, i64* %7, i64 %6
  %9 = load i64, i64* %8, align 8
  %10 = trunc i64 %9 to i8
  %11 = add i8 65, %10
  %12 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([4 x i8], [4 x i8]* @S0, i32 0, i32 0), i8 %11)
  br label %13

13:                                               ; preds = %5
  %14 = load i64, i64* %3, align 8
  %15 = add i64 %14, 1
  store i64 %15, i64* %3, align 8
  br label %17

16:                                               ; preds = %17
  ret i32 0

17:                                               ; preds = %13, %1
  %18 = load i64, i64* %3, align 8
  %19 = icmp ult i64 %18, 8
  br i1 %19, label %5, label %16

20:                                               ; preds = %0
  br label %1
}

attributes #0 = { nounwind }
```

`error.ll`
```llvm
source_filename = ""
target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128"
target triple = "arm64-apple-darwin21.4.0"

@S0 = internal constant [4 x i8] c"%c\0A\00"

declare i32 @printf(i8* %0, ...)

; Function Attrs: nounwind
define private [9 x i64] @a41() #0 {
  %1 = alloca [9 x i64], align 8
  %2 = alloca i64, align 8
  store [9 x i64] zeroinitializer, [9 x i64]* %1, align 8
  store i64 0, i64* %2, align 8
  br label %13

3:                                                ; preds = %13
  %4 = load i64, i64* %2, align 8
  %5 = load i64, i64* %2, align 8
  %6 = bitcast [9 x i64]* %1 to i64*
  %7 = getelementptr i64, i64* %6, i64 %5
  store i64 %4, i64* %7, align 8
  br label %8

8:                                                ; preds = %3
  %9 = load i64, i64* %2, align 8
  %10 = add i64 %9, 1
  store i64 %10, i64* %2, align 8
  br label %13

11:                                               ; preds = %13
  %12 = load [9 x i64], [9 x i64]* %1, align 8
  ret [9 x i64] %12

13:                                               ; preds = %8, %0
  %14 = load i64, i64* %2, align 8
  %15 = icmp ult i64 %14, 9
  br i1 %15, label %3, label %11
}

; Function Attrs: nounwind
define i32 @main() #0 {
  br label %20

1:                                                ; preds = %20
  %2 = alloca [9 x i64], align 8
  %3 = alloca i64, align 8
  %4 = call [9 x i64] @a41()
  store [9 x i64] %4, [9 x i64]* %2, align 8
  store i64 0, i64* %3, align 8
  br label %17

5:                                                ; preds = %17
  %6 = load i64, i64* %3, align 8
  %7 = bitcast [9 x i64]* %2 to i64*
  %8 = getelementptr i64, i64* %7, i64 %6
  %9 = load i64, i64* %8, align 8
  %10 = trunc i64 %9 to i8
  %11 = add i8 65, %10
  %12 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([4 x i8], [4 x i8]* @S0, i32 0, i32 0), i8 %11)
  br label %13

13:                                               ; preds = %5
  %14 = load i64, i64* %3, align 8
  %15 = add i64 %14, 1
  store i64 %15, i64* %3, align 8
  br label %17

16:                                               ; preds = %17
  ret i32 0

17:                                               ; preds = %13, %1
  %18 = load i64, i64* %3, align 8
  %19 = icmp ult i64 %18, 9
  br i1 %19, label %5, label %16

20:                                               ; preds = %0
  br label %1
}

attributes #0 = { nounwind }
```

**Comment by aemerson** [comment]
It seems we never implemented support for sret demotion properly on AArch64 GlobalISel. Matt implemented it for AMDGPU when he added support but we have a missing piece for the `canLowerReturn()` override. Not that this is his fault, sret support was broken before anyway.

**Comment by aemerson** [comment]
Patch to fix this: https://reviews.llvm.org/D129098

**Comment by aemerson** [comment]
Fixed in [b97013fd6097](https://reviews.llvm.org/rGb97013fd60977ba62a747729183e033c0900de8c)
